File: basic-go/leetcode/list/rotate_right.go
package list

// 给定一个链表，将链表的每个节点向右移动 k 个位置

// 找到倒数 k 个节点，断开链接，然后将尾节点指向头节点
// 使用双指针法，使两个指针的间隔为 k，当快指针到达链表尾部时慢指针到达倒数 k 个节点
// 时间复杂度 O(N)
func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil {
		return head
	}

	// 计算单链表长度
	length := 0
	for curr := head; head != nil; curr = curr.Next {
		length++
	}

	fast, slow := head, head

	// 取模防止 k > length
	for fast != nil && k%length > 0 {
		fast, k = fast.Next, k-1
	}

	// fast 指向尾节点
	for fast.Next != nil {
		slow, fast = slow.Next, fast.Next
	}

	fast.Next = head
	head = slow.Next
	slow.Next = nil
	return head

}

// 将链表转换成环形链表，然后找到倒数 k 位置断开链接
// k = n - (k % n)
func rotateRight1(head *ListNode, k int) *ListNode {

	if head == nil || head.Next == nil {
		return head
	}

	curr := head
	n := 1
	// 转换单链表为环形链表，并且计算链表长度
	for curr.Next != nil {
		curr, n = curr.Next, n+1
	}
	curr.Next = head

	curr = head
	// 找到倒数 k 位置节点断开链接
	for i := 0; i < n-k%n-1; i++ {
		curr = curr.Next
	}
	head := curr.Next
	curr.Next = nil
	return head
}

File: basic-algorithm/src/example/leetcode/list/RotateRight.java
package example.leetcode.list;

/**
 * 给定链表头节点 head，将链表节点向右移动 K 个位置
 *
 */
public class RotateRight {

    public ListNode rotateRight(ListNode head, int k){

        /*
         * 注意 K 有可能会大于链表长度
         */
        int length = 0;
        ListNode p = head;
        while (p != null){
            length++;
            p = p.next;
        }
        if(length == 0){
            return head;
        }
        if (k > length){
            k = k % length;
        }
        if(k == 0){
            return head;
        }

        ListNode dummy = new ListNode(-1, head);

        // 找到倒数 K 个节点，断开连接，将尾结点和头结点连接
        ListNode left = dummy, right = dummy;
        int i = 0;
        while (right.next != null){
            right = right.next;
            if (i < k){
                i++;
            }else{
                left = left.next;
            }
        }
        // 不足 k 个
        if (i < k){
            return dummy.next;
        }
        ListNode tmp = left.next;
        left.next = null;
        right.next = dummy.next;
        dummy.next = tmp;

        return dummy.next;
    }


    /**
     * 构造成环形链表
     */
    public ListNode rotateRight_1(ListNode head, int k){

        if (head == null){
            return null;
        }
        ListNode curr = head;
        int length = 1;
        while (curr.next != null){
            curr = curr.next;
            length++;
        }
        curr.next = head;

        k = k % length;
        int move = length - k - 1;
        while (move > 0){
            head = head.next;
            move--;
        }
        ListNode result = head.next;
        head.next = null;
        return result;
    }
}
